Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp

```diff
@@ -0,0 +1,257 @@
+
+    return dummyURL.protocol().toString();
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-a-username, combined with https://urlpattern.spec.whatwg.org/#process-username-for-init
+String canonicalizeUsername(const String& value, BaseURLStringType valueType)
+{
+    if (value.isEmpty())
+        return String { value };
+
+    if (valueType == BaseURLStringType::Pattern)
+        return String { value };
+
+    URL dummyURL("https://www.webkit.org"_s);
+    dummyURL.setUser(value);
+
+    return dummyURL.user();
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-a-password, combined with https://urlpattern.spec.whatwg.org/#process-password-for-init
+String canonicalizePassword(const String& value, BaseURLStringType valueType)
+{
+    if (value.isEmpty())
+        return String { value };
+
+    if (valueType == BaseURLStringType::Pattern)
+        return String { value };
+
+    URL dummyURL("https://www.webkit.org"_s);
+    dummyURL.setPassword(value);
+
+    return dummyURL.password();
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-a-hostname, combined with https://urlpattern.spec.whatwg.org/#process-hostname-for-init
+ExceptionOr<String> canonicalizeHost(const String& value, BaseURLStringType valueType)
+{
+    if (value.isEmpty())
+        return String { value };
+
+    if (valueType == BaseURLStringType::Pattern)
+        return String { value };
+
+    URL dummyURL("https://www.webkit.org"_s);
+    dummyURL.setHost(value);
+
+    if (!dummyURL.isValid())
+        return Exception { ExceptionCode::TypeError, "Invalid input to canonicalize a URL host string."_s };
+
+    return String { value };
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-an-ipv6-hostname
+ExceptionOr<String> canonicalizeIPv6Host(const String& value, BaseURLStringType valueType)
+{
+    if (valueType == BaseURLStringType::Pattern)
+        return String { value };
+
+    StringBuilder result;
+    result.reserveCapacity(result.length());
+
+    for (auto codepoint : StringView(value).codePoints()) {
+        if (isInvalidIPv6HostCodePoint(codepoint))
+            return Exception { ExceptionCode::TypeError, "Invalid input to canonicalize a URL IPv6 host string."_s };
+
+        result.append(toASCIILower(codepoint));
+    }
+
+    return result.toString();
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-a-port, combined with https://urlpattern.spec.whatwg.org/#process-port-for-init
+ExceptionOr<String> canonicalizePort(const String& portValue, std::optional<StringView> protocolValue, BaseURLStringType portValueType)
+{
+    if (portValue.isEmpty())
+        return String { portValue };
+
+    if (portValueType == BaseURLStringType::Pattern)
+        return String { portValue };
+
+    URL dummyURL("https://www.webkit.org"_s);
+
+    if (protocolValue.has_value())
+        dummyURL.setProtocol(protocolValue.value());
```

Comment:
No big deal but this is a bit shorter `*protocolValue` and you have used it earlier in your PR.